rust/rope/src/compare.rs

```diff
@@ -110,38 +113,27 @@ pub fn ne_idx_avx(one: &[u8], two: &[u8]) -> Option<usize> {
 }
 
 #[cfg(target_arch = "x86_64")]
-#[allow(dead_code)]
+#[target_feature(enable="sse4.2")]
 #[doc(hidden)]
-//NOTE: this turned out to be slower than sw, keeping it around
-//for now for benchmarking.
-pub fn ne_idx_sse42(one: &[u8], two: &[u8]) -> Option<usize> {
+pub unsafe fn ne_idx_sse(one: &[u8], two: &[u8]) -> Option<usize> {
     let min_len = one.len().min(two.len());
     let mut idx = 0;
-    loop {
-        let mask: i32;
-        // if slice is less than 16 bytes we manually pad it
-        if idx + SSE_STRIDE >= min_len {
-            let mut one_buf: [u8; SSE_STRIDE] = [0; SSE_STRIDE];
-            let mut two_buf: [u8; SSE_STRIDE] = [0; SSE_STRIDE];
-            one_buf[..min_len-idx].copy_from_slice(&one[idx..min_len]);
-            two_buf[..min_len-idx].copy_from_slice(&two[idx..min_len]);
-            mask = unsafe { fast_cmpestr_mask(&one_buf, &two_buf) };
-        } else {
-            mask = unsafe { fast_cmpestr_mask(&one[idx..idx+SSE_STRIDE],
-                                              &two[idx..idx+SSE_STRIDE]) };
+    while idx < min_len {
+        let stride_len = SSE_STRIDE.min(min_len - idx);
+        let mask = sse_compare_mask(&one.get_unchecked(idx..idx+stride_len),
+                                    &two.get_unchecked(idx..idx+stride_len));
+        if mask != 0 && idx + (mask.trailing_zeros() as usize) < min_len {
+            return Some(idx + mask.trailing_zeros() as usize);
         }
-        let i = mask.trailing_zeros() as usize;
-        if i != 32 { return Some(idx + i); }
         idx += SSE_STRIDE;
-        if idx >= min_len { break; }
     }
     None
 }
 
 #[cfg(target_arch = "x86_64")]
```

Comment:
This definitely needs to be marked unsafe, right now it's a safety hole. Also it should be `#[target_feature(enable)]`. My sense is that you made changes to the forward version and didn't make corresponding changes for rev?